File: examples/canonical-auth/java/src/main/java/dev/tenantcore/auth/routes/PermissionController.java
package dev.tenantcore.auth.routes;

import dev.tenantcore.auth.domain.InMemoryStore;
import dev.tenantcore.auth.domain.Permission;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Collection;

@RestController
public class PermissionController {

    private final InMemoryStore store;

    public PermissionController(InMemoryStore store) {
        this.store = store;
    }

    @GetMapping("/permissions")
    @PreAuthorize("hasAuthority('iam:permission:read')")
    public ResponseEntity<Collection<Permission>> listPermissions() {
        return ResponseEntity.ok(store.getAllPermissions());
    }
}

File: examples/canonical-auth/elixir/lib/tenantcore_auth/controllers/permission_controller.ex
defmodule TenantcoreAuth.PermissionController do
  use TenantcoreAuth, :controller

  plug TenantcoreAuth.Auth.RbacPlug, "iam:permission:read"

  def index(conn, _params) do
    permissions =
      TenantcoreAuth.Domain.InMemoryStore.list_permissions()
      |> Enum.map(fn permission -> %{id: permission.id, name: permission.name, description: permission.description} end)

    json(conn, %{permissions: permissions})
  end
end
